crates/pathfinder/src/state/sync/l2.rs

```diff
@@ -471,21 +506,58 @@ async fn download_and_compress_class(
         class_hash.0
     );
 
-    let compress = tokio::task::spawn_blocking(move || -> anyhow::Result<_> {
-        let mut compressor = zstd::bulk::Compressor::new(10).context("Create zstd compressor")?;
+    match hash {
+        starknet_gateway_types::class_hash::ComputedClassHash::Cairo(hash) => {
+            let compress = tokio::task::spawn_blocking(move || -> anyhow::Result<_> {
+                let mut compressor =
+                    zstd::bulk::Compressor::new(10).context("Create zstd compressor")?;
 
-        let definition = compressor
-            .compress(&definition)
-            .context("Compress definition")?;
+                let definition = compressor
+                    .compress(&definition)
+                    .context("Compress definition")?;
 
-        Ok(definition)
-    });
-    let definition = compress.await.context("Compress contract")??;
+                Ok(definition)
+            });
+            let compressed_definition = compress.await.context("Compress contract")??;
 
-    Ok(CompressedContract {
-        definition,
-        hash: hash.hash(),
-    })
+            Ok(DownloadedClass::Cairo(CompressedContract {
+                definition: compressed_definition,
+                hash,
+            }))
+        }
+        starknet_gateway_types::class_hash::ComputedClassHash::Sierra(hash) => {
+            let casm_definition =
+                crate::sierra::compile_to_casm(&definition).context("Compiling Sierra class")?;
+
+            let compress = tokio::task::spawn_blocking(move || -> anyhow::Result<_> {
```

Comment:
Wonder if we shouldn't consider creating a `async fn blocking_(de)compress` since we do this pattern so often.

Reply:
We could -- but then compressing _two_ things (like Sierra & CASM definition) would be two tasks, right? I think that would cause some overhead?

Reply:
Ah I see yes. Hmm. Maybe we can do a nice builder pattern for all things we consider blocking tasks? Although that might be more trouble than its worth..

```rust
let task_result = BlockingTasks::builder()
    .compress(sierra_definition)
    .compress(casm_definition)
    .other_task_we_have(inputs)
    .spawn_blocking();
```